README

= classx
Meta Framework extending and flexible attribute like Moose ( perl )

== Synopsis

  require 'classx'
  class Point
    include ClassX

    has :x, :kind_of => Fixnum
    has :y, :kind_of => Fixnum
  end

  class Point3D < Point
    has :z, :kind_of => Fixnum, :optional => true
  end

  Point.new(:x => 30, :y => 40)  #=> <# Point @x=30, @y=40 >
  point3d = Point3D.new(:x => 30, :y => 40, :z => 50)  #=> <# Point3D @x=30, @y=40, @z=50 >
  point3d.z = 60.0 # raise ClassX::InvalidAttrArgument

== Description
perl's Moose like Class interface.

== Installation

=== Archive Installation

	rake install

=== Gem Installation

	gem install classx

== Features/Problems

== SEE ALSO

Moose: http://search.cpan.org/~drolsky/Moose-0.57/lib/Moose.pm

== Copyright

Author::    Keiji, Yoshimi <walf443 at gmail.com>
Copyright:: Copyright (c) 2008 Keiji, Yoshimi
License:: you can redistribute it and/or modify it under the same terms as Ruby itself.


ChangeLog
commit bb9c1d8e13ceb2b46c0ae5a002fad4f5f2d57230
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Mon Feb 16 22:37:45 2009 +0900

    fixed bug: when false was set as default value, getter return nil. thanks niku!!

 lib/classx.rb                      |    2 +-
 lib/classx/attributes.rb           |    8 ++++----
 spec/classx/default_option_spec.rb |   32 ++++++++++++++++++++++++++++++++
 3 files changed, 37 insertions(+), 5 deletions(-)

commit 1cd4943cf71c96d09d6aef15afaf79ec86cd0b5b
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Dec 26 09:16:02 2008 +0900

    release 0.0.6

 ChangeLog |  560 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 Rakefile  |    2 +-
 2 files changed, 556 insertions(+), 6 deletions(-)

commit 1bc6ca2722a0bfe96ae3f7f9af76a5e30f1872ac
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Dec 26 09:15:08 2008 +0900

    add some doc.

 README                   |    2 +-
 lib/classx.rb            |    2 +-
 lib/classx/attributes.rb |    5 ++++-
 3 files changed, 6 insertions(+), 3 deletions(-)

commit 2a1f45c859f24ab2757fdb2746d7e7a0cb8b6a5c
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Dec 26 08:53:42 2008 +0900

    CAVEAT: change default option to :writable is true.
    
    If you want to non-rewritable attribute, you should declare :writable option
    explicitly.

 lib/classx/attributes.rb                           |    2 +-
 lib/classx/class_attributes.rb                     |    2 +-
 spec/classx/without_anyoption_spec.rb              |    6 ++++--
 .../without_anyoption_spec.rb                      |    5 +++--
 .../writable_option_spec.rb                        |    2 +-
 5 files changed, 10 insertions(+), 7 deletions(-)

commit be536d9561443f9fdb2cfa360f244d6ff0746b89
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Dec 26 00:52:37 2008 +0900

    FIXED: :coerce feature not work when you rewrite attribute data. thanks niku!!

 lib/classx/attributes.rb        |    3 +--
 spec/classx/with_coerce_spec.rb |    8 +++++++-
 2 files changed, 8 insertions(+), 3 deletions(-)

commit 73745a8ef87eed4be130602c78c8407688c8076b
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Sat Dec 20 12:04:32 2008 +0900

    change web site url to github rdoc page.

 tasks/basic_config.rake |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 44c6ff607e67c6ff52670f0d15eb84fa200b1a90
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Sat Dec 20 11:52:42 2008 +0900

    using rdoc template allison.

 tasks/basic_tasks.rake |   31 ++++++++++++++++++-------------
 1 files changed, 18 insertions(+), 13 deletions(-)

commit 21d52770eefaf66d7599dde2389bb12b6108357d
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Tue Nov 4 00:47:37 2008 +0900

    tweaked error message.

 lib/classx/attribute.rb |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit e01d94a5070de17b012b26d10b08f3b24ad23c59
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Tue Nov 4 00:02:46 2008 +0900

    added :options :cache_key to ClassX::Validate. So you can turn off caching anonymous validate class feature.

 lib/classx/validate.rb       |   29 +++++++++++++++++------
 spec/classx_validate_spec.rb |   52 +++++++++++++++++++++++++++++++----------
 2 files changed, 60 insertions(+), 21 deletions(-)

commit 17a7717af41d9c55ff3b43a21bab445c3f78d2eb
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Mon Oct 27 00:03:47 2008 +0900

    release 0.0.5

 ChangeLog |  114 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Rakefile  |    2 +-
 2 files changed, 115 insertions(+), 1 deletions(-)

commit de3b716466a8c30e31f75483977bfcd9f19f871a
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Mon Oct 27 00:00:14 2008 +0900

    oops. spec message fixed.

 spec/classx/handles_spec.rb |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

commit 8c892fc051bba7175cbaa02dd9d29cdd3bab7d51
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Sun Oct 26 23:00:20 2008 +0900

    update benchmark result.

 bench/define_attribute.rb |   18 ++++++++++++++++++
 1 files changed, 18 insertions(+), 0 deletions(-)

commit ad6a1e8be27e44d38b2a357f060086053572ecb2
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Sun Oct 26 18:53:28 2008 +0900

    fixed bug. when you handle some methods that use with block, it not worked.

 lib/classx/attributes.rb    |   16 ++++++++++------
 spec/classx/handles_spec.rb |   15 +++++++++++++--
 2 files changed, 23 insertions(+), 8 deletions(-)

commit ad34cf6a7f3d119379a2bc78b4060504b404db9e
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Sep 19 08:59:34 2008 +0900

    added coerce to log_rotate.

 lib/classx/role/logger.rb |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

commit 8b4cc50e9395bb801beafd60afc579e279e7604d
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Sep 19 08:52:38 2008 +0900

    added CoerceWithProc.

 lib/classx/attribute.rb                          |    8 ++++
 spec/classx/with_coerce_spec.rb                  |   20 ++++++++++
 spec/classx_class_attributes/with_coerce_spec.rb |   43 ++++++++++++++++++----
 3 files changed, 63 insertions(+), 8 deletions(-)

commit cc2917b5b7f1088428936f9065932c7d630ae356
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Sep 19 08:37:39 2008 +0900

    added validate to log_rotate.

 lib/classx/role/logger.rb |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

commit ef422314b28c741950a04e4a9d75a532c4f4a0ab
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Sep 19 08:29:22 2008 +0900

    fixed bug:  ClassX::Commandable with none kind_of specified attribute does not output :desc to help.

 lib/classx/commandable.rb |    4 ++--
 lib/classx/role/logger.rb |    2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

commit 62217b9347f6470db829ef66de116a9bdab04c7c
Author: Keiji, Yoshimi <walf443@gmail.com>
Date:   Fri Sep 19 08:20:01 2008 +0900

    added more shortcut.

 example/commandable.rb    |    6 +++---

lib/classx.rb
require 'classx/attribute_factory'
require 'classx/attributes'

# usage
#
#   require 'classx'
#   class Point
#     include ClassX
#   
#     has :x, :kind_of => Fixnum
#     has :y, :kind_of => Fixnum
#   end
#   
#   class Point3D < Point
#     has :z, :kind_of => Fixnum, :optional => true
#   end
#   
#   Point.new(:x => 30, :y => 40)  #=> <# Point @x=30, @y=40 >
#   point3d = Point3D.new(:x => 30, :y => 40, :z => 50)  #=> <# Point3D @x=30, @y=40, @z=50 >
#   point3d.z = 60.0 # raise ClassX::InvalidAttrArgument
module ClassX
  autoload :ClassAttributes, 'classx/class_attributes'
  autoload :CAttrs,          'classx/class_attributes'
  autoload :Validate,    'classx/validate'
  autoload :Commandable, 'classx/commandable'
  autoload :Declare,     'classx/declare'
  autoload :Bracketable, 'classx/bracketable'
  module Role
    autoload :Logger,    'classx/role/logger'
  end

  MODULE_USAGE_MAP_OF = {
    :ClassAttributes => :extend,
    :CAttrs => :extend,
    :Commandable => :extend,
    :Declare     => :extend,
    :Bracketable => :include,
    :Validate => :include,
  }

  class InstanceException < Exception; end
  class AttrRequiredError < InstanceException; end
  class InvalidAttrArgument < InstanceException; end
  class LazyOptionShouldHaveDefault < Exception; end
  class OptionalAttrShouldBeWritable < Exception; end
  class RequiredAttrShouldNotHaveDefault < Exception; end

  # *args is Hash in default.
  # Hash should have attribute's key and valid value for attribute.
  # This method checking required value is setted and taking value is valid to attribute.
  #
  def initialize *args
    if respond_to? :before_init
      warn 'before_init is deprecated. please define process_init_args class method'
    else
      hash = self.class.process_init_args(*args)
    end

    unless hash && hash.kind_of?(Hash)
      raise ArgumentError, "#{hash.inspect} was wrong as arguments. please specify kind of Hash instance"
    end

    # allow String or Symbol for key
    tmp_hash = {}
    hash.each do |key,val|
      tmp_hash[key.to_s] = val
    end
    hash = tmp_hash

    hash.each do |key, val|
      if attribute_of[key]
        attribute_of[key].set val
      end
    end

    attribute_of.each do |key, val|
      next if val.class.lazy?
      raise AttrRequiredError, "param: :#{key} is required to #{hash.inspect}" if !val.class.optional? && val.get.nil?
    end

    after_init
  end

  # return Hash of attribute's name as a key and attribute's meta class instance as a value.
  # for example, 
  #
  #   class YourClass
  #      include ClassX
  #      has :x
  #   end
  #
  #   obj = YourClass.new(:x => 10)
  #   obj.attribute_of    #=> { "x" => <#<ClassX::Attribute> parent=<# YourClass> ... > }
  #
  def attribute_of
    unless instance_variable_defined?('@__attribute_of') && @__attribute_of
      @__attribute_of = {}
      if self.class.attribute_of
        self.class.attribute_of.keys.each do |key|
          @__attribute_of[key] = __send__ "attribute_of:#{key}"
        end
      end
    end

    @__attribute_of
  end

  # automatically called this method on last of #initialize.
  # you can override this method.
  def after_init
  end

  # shared implementation for comparing classx based object.
  def == other
    return false unless other.kind_of? self.class
    attribute_of.all? do |key, val|
      val.get == other.__send__(key)
    end
  end

  UNSERIALIZE_INSTANCE_VARIABLES = ["@__attribute_of", "@__attribute_data_of"]

  # convert attribute key and value to Hash.
  def to_hash
    result = {}

    attribute_of.each do |key, val|
      result[key] = val.get
    end

    result
  end

  # dupping classx based class.
  def dup
    self.class.new(to_hash)
  end

  # for Marshal.dump
  def marshal_dump
    dump_of = {}
    dump_of[:attribute_of] = to_hash
    dump_of[:instance_variable_of] = {}
    ( instance_variables.map {|ival| ival.to_s } - UNSERIALIZE_INSTANCE_VARIABLES ).each do |ival|
      dump_of[:instance_variable_of][ival] = instance_variable_get(ival)
    end

    dump_of
  end

  # for Marshal.load
  def marshal_load val
    self.attribute_of.each do |k, v|
      v.set(val[:attribute_of][k])
    end
    val[:instance_variable_of].each do |key, val|
      instance_variable_set(key, val)
    end
  end

  # for YAML.dump
  def to_yaml opts={}
    require 'yaml'
    YAML.quick_emit(self, opts) do |out|
      out.map( taguri, to_yaml_style ) do |map|
        attribute_of = {}
        to_hash.each do |key, val|
          attribute_of[key] = val
        end
        map.add(:attribute_of, attribute_of)

        instance_variable_of = {}
        ( instance_variables.map {|ival| ival.to_s } - UNSERIALIZE_INSTANCE_VARIABLES ).each do |ival|
          instance_variable_of[ival] = instance_variable_get(ival)
        end
        map.add(:instance_variable_of, instance_variable_of)
      end
    end
  end

  # for YAML.load
  def yaml_initialize tag, val
    self.attribute_of.each do |k, v|
      v.set(val[:attribute_of][k])
    end

    val[:instance_variable_of].each do |k, v|
      instance_variable_set(k, v)
    end
  end

  module ClassMethods
    # processing initialize argument to hash
    # you can override this method for not taking initializer your classx based class.
    def process_init_args *args
      raise ArgumentError if args.size > 1

      hash = args.first
      hash.nil? ? {} : hash
    end
  end

  private

  def self.included klass
    klass.extend(Attributes)
    klass.extend(ClassMethods)
  end

end
